controlplane/kubeadm/internal/cluster.go

```diff
@@ -0,0 +1,337 @@
+		EtcdCACert:etcdCACert,
+		EtcdCAKey:etcdCAKey,
+	}, nil
+}
+
+// GetEtcdCA returns the EtcdCA Cert and Key for a given cluster.
+func (m *ManagementCluster) GetEtcdCerts(ctx context.Context, cluster types.NamespacedName) ([]byte, []byte, error) {
+	etcdCASecret := &corev1.Secret{}
+	etcdCAObjectKey := types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      fmt.Sprintf("%s-etcd", cluster.Name),
+	}
+	if err := m.Client.Get(ctx, etcdCAObjectKey, etcdCASecret); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to get secret; etcd CA bundle %s/%s", etcdCAObjectKey.Namespace, etcdCAObjectKey.Name)
+	}
+	return etcdCASecret.Data["tls.crt"], etcdCASecret.Data["tls.key"], nil
+}
+
+func a() {
+
+}
+
+// ControlPlaneLabels returns a set of labels to add to a control plane machine for this specific cluster.
+func (m *ManagementCluster) ControlPlaneLabelsForCluster(clusterName string) map[string]string {
+	return map[string]string{
+		clusterv1.ClusterLabelName:             clusterName,
+		clusterv1.MachineControlPlaneLabelName: "",
+	}
+}
+
+// ControlPlaneSelectorForCluster returns the label selector necessary to get control plane machines for a given cluster.
+func (m *ManagementCluster) ControlPlaneSelectorForCluster(clusterName string) *metav1.LabelSelector {
+	return &metav1.LabelSelector{
+		MatchLabels: m.ControlPlaneLabelsForCluster(clusterName),
+	}
+}
+
+// Cluster are operations on target clusters
+type Cluster struct {
+	Client ctrlclient.Client
+	// RestConfig is required for the proxy.
+	RestConfig *rest.Config
+	EtcdCACert, EtcdCAKey []byte
+}
+
+// generateEtcdTLSClientBundle builds an etcd client TLS bundle from the Etcd CA for this cluster.
+func (c *Cluster) generateEtcdTLSClientBundle() (*tls.Config, error) {
+	clientCert, err := generateClientCert(c.EtcdCACert,c.EtcdCAKey)
+	if err != nil {
+		return nil, err
+	}
+
+	caPool := x509.NewCertPool()
+	caPool.AppendCertsFromPEM(c.EtcdCACert)
+
+	return &tls.Config{
+		RootCAs:      caPool,
+		Certificates: []tls.Certificate{clientCert},
+	}, nil
+}
+
+// EtcdIsHealthy runs checks for every etcd member in the cluster to satisfy our definition of healthy.
+// This is a best effort check and nodes can become unhealthy after the check is complete. It is not a guarantee.
+// It's used a signal for if we should allow a target cluster to scale up, scale down or upgrade.
+func (c *Cluster) EtcdIsHealthy(ctx context.Context) error {
+	var errs []error
+	var knownClusterID uint64
+	var knownMemberIDSet etcdutil.UInt64
+
+	controlPlaneNodes := &corev1.NodeList{}
+	controlPlaneNodeLabels := map[string]string{
+		"node-role.kubernetes.io/master": "",
+	}
+
+	if err := c.Client.List(ctx, controlPlaneNodes, client.MatchingLabels(controlPlaneNodeLabels)); err != nil {
+		return err
+	}
+
+	tlsConfig, err := c.generateEtcdTLSClientBundle()
+	if err != nil {
+		return err
+	}
+
+	// TODO: Do we need a way to see if there is a 1:1 correspondence of Machines to Nodes?
```

Comment:
This is the major open question here: Is it ok to delegate the responsibility of checking for etcd health to cluster itself?

The difference is the target cluster won't have any notion of machines as they are stored on the management cluster.

Allowing the cluster to tell us if it thinks etcd is healthy *without* confirming that the nodes on the target cluster match the machines known in cluster-api is a bigger foot-gun than if we did that check, but it also allows for more use cases (however unsupported they may be). For example, if a user manually joins a control plane node to a target cluster there would be an additional node unmanaged by a machine. In the current version of this code, that would be acceptable so long as it's healthy. Cluster API would happily scale up. If that is not the behavior we want then we should return a list of nodes that were checked and allow something else (mgmt or reconciler) to check and see if the list of nodes matches the list of machines.

Also this might be an edge case that we can not care about for now and fix it later if it becomes a problem. I'll open an issue if we want to move forward with that approach.

Reply:
I think the Cluster API Machine resources must be the reconciler's source of truth for desired state. I consider the target cluster's Node resources to be actual state.

For that reason, I would prefer the reconciler to work with list of Machines its owns, as opposed to working with a list of Nodes.

What if we replace the "target cluster" abstraction with a "target control plane" one? The reconciler can initialize this "target control plane" with a list of Machines, and the abstraction's implementation can map Machines to Nodes when necessary .

That is, instead of
```
type Cluster struct {
	Client ctrlclient.Client
	// RestConfig is required for the proxy.
	RestConfig *rest.Config
	EtcdCACert, EtcdCAKey []byte
}
```

We have
```
type ControlPlane struct {
	Machines clusterv1.MachineList
	Client ctrlclient.Client
	// RestConfig is required for the proxy.
	RestConfig *rest.Config
	EtcdCACert, EtcdCAKey []byte
}
```

Reply:
How would you feel about the alternative solution we discussed where the health check reports on the nodes it checked and something else compares that result against the known list of machines? That feels cleaner to me and allows Cluster to not know anything at all about Cluster API

Reply:
I'm ok with that.

Something, somewhere should verify that control plane Machines and control plane Nodes have 1:1 correspondence.